qemu/unicorn_common.h

```diff
@@ -28,6 +29,68 @@ static inline bool cpu_physical_mem_write(AddressSpace *as, hwaddr addr,
     return cpu_physical_memory_rw(as, addr, (void *)buf, len, 1);
 }
 
+static bool cpu_virtual_mem_read(struct uc_struct *uc, vaddr addr, uint32_t prot, uint8_t *buf, int len)
+{
+    MMUAccessType access_type;
+    void *hostptr;
+    int mmu_idx = cpu_mmu_index(uc->cpu->env_ptr, false);
+
+    /*
+     * Only page aligned access is allowed,
+     * because tlb_fill() might change the mappings
+     */
+    assert((addr & TARGET_PAGE_MASK) == ((addr + len - 1) & TARGET_PAGE_MASK));
```

Comment:
this shall return `UC_ERR_ARG` or more appropriate errors instead of assertions.

Reply:
I use asserts for potential unicorn bugs and return errors for user bugs. This internal function is designed to only be called within one page. This is implemented in `uc_mem_read_virtual`. Returning an error might cause a user to believe it's a bug in there code.

I'll write a comment about this.

Reply:
Yeah, but addr is from user inputs, no? `uc_mem_read_virtual` seems using it directly? My concern is that this might cause the assertion too easy to trigger.

Reply:
Yes addr is (more or less) from user input, but len is enforced to be maximum the rest of the page. What this assert checks is that the full len is on the same page. So the assert is in exactly what `uc_mem_read_virtual` enforces. When we add other callers, they must handle this internal api the same way.

So I don't see how this assert triggers to early.